src/Microsoft.Framework.Runtime/ApplicationHostContext.cs

```diff
@@ -65,34 +76,30 @@ public class ApplicationHostContext
             {
                 var lockFileReader = new LockFileReader();
                 var lockFile = lockFileReader.Read(projectLockJsonPath);
-                validLockFile = lockFile.IsValidForProject(Project);
 
-                if (validLockFile || skipLockFileValidation)
+                string message;
+                validLockFile = lockFile.IsValidForProject(Project, out message);
+
+                if (skipLockFileValidation || validLockFile)
                 {
                     NuGetDependencyProvider.ApplyLockFile(lockFile);
-
-                    DependencyWalker = new DependencyWalker(new IDependencyProvider[] {
-                        ProjectDepencyProvider,
-                        NuGetDependencyProvider,
-                        referenceAssemblyDependencyResolver,
-                        gacDependencyResolver,
-                        unresolvedDependencyProvider
-                    });
+                }
+                else
+                {
+                    // Don't add NuGetDependencyProvider to DependencyWalker
+                    dependencyProviders = dependencyProviders.Where(d => d != NuGetDependencyProvider);
+                    _diagnostics.Add(new LockFileValidationMessage(message, projectLockJsonPath));
                 }
             }
-
-            if ((!validLockFile && !skipLockFileValidation) || !lockFileExists)
+            else
             {
-                // We don't add NuGetDependencyProvider to DependencyWalker
-                // It will leave all NuGet packages unresolved and give error message asking users to run "dnu restore"
-                DependencyWalker = new DependencyWalker(new IDependencyProvider[] {
-                    ProjectDepencyProvider,
-                    referenceAssemblyDependencyResolver,
-                    gacDependencyResolver,
-                    unresolvedDependencyProvider
-                });
+                // Don't add NuGetDependencyProvider to DependencyWalker
```

Comment:
This is comment does not add any value in its current form. It's quite obvious from the line below that you are not adding NuGetDependencyProvider but the answer to the "why" is missing. Same above.